---
title: Promise
date: 2022-11-23
tag:
  - ES6
  - JavaScript
categories:
  - JavaScript
---

## Promise

### 基本用法

`Promise` 是类或构造函数，通过对它实例化，来完成预期的异步任务处理。

`Promise` 有三种状态：
- pending（初始状态）
- fulfilled（成功）
- rejected（失败）

```JavaScript
const p = new Promise((resolve, reject) => {
  // 异步任务主体部分
  resolve('success') // 或 reject('failed')
})
```

通过 resolve() 将状态修改为 fulfilled ，resolve(data) 可以返回数据，在 then() 回调里处理。

```JavaScript
p.then(data => {
  // 数据处理
})
```

### Promise 的异常处理

**方法一：通过 then() 的第二个参数**

通过 reject() 将状态修改为 rejected， reject(error) 可以返回错误，在 then() 第二个参数回调里处理。

```JavaScript
p.then(data => {
  // 数据处理
}, error => {
  // 错误处理
})
```

这种方式可以捕获 `Promise` 主体代码的错误，但是如果主体代码没有错误，然后错误发生在 then() 的成功回调，那么这种方式无法捕获错误。

**方法二：通过 catch() 处理**

```JavaScript
p.then(data => {
  // 数据处理
}).catch(error => {
  // 错误处理
})
```

这种方式无论是 `Promise` 主体代码的错误，还是在成功回调里发生的错误，都可以捕获。

**方法三：使用 try...catch**

try catch 是传统的异常捕获方式，不过只能捕获同步代码的异常，无法捕获异步代码的异常。
